### PR TITLE
[NFC] Make the GCData constructor a move constructor

### PR DESCRIPTION
**Pull request by kripken**
This avoids creating a large Literals (SmallVector of Literal) and then copying it. All the places
that construct GCData do not need the Literals afterwards.

Also add a few other constructor type things, all as a result of C++ compiler errors after
the GCData constructor change.

This gives a 7% speedup on the `--precompute` benchmark from #6931

### CODE REVIEW

File: src/support/small_vector.h [hunk 1]
```diff
@@ -65,13 +65,26 @@ template<typename T, size_t N> class SmallVector {
...
+    : usedFixed(other.usedFixed), fixed(std::move(other.fixed)),
+      flexible(std::move(other.flexible)) {}
   SmallVector(std::initializer_list<T> init) {
     for (T item : init) {
       push_back(item);
     }
   }
   SmallVector(size_t initialSize) { resize(initialSize); }
 
+  SmallVector<T, N>& operator=(const SmallVector<T, N>& other) {
```
Comment (tlively) [inline src/support/small_vector.h:81]: Should we add a move assignment operator for completeness?

File: src/wasm-interpreter.h [hunk 2]
```diff
@@ -191,8 +191,11 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
   // data, as we don't have a cycle collector. Those leaks are not a serious
   // problem as Binaryen is not really used in long-running tasks, so we ignore
   // this function in LSan.
-  Literal makeGCData(const Literals& data, Type type) {
-    auto allocation = std::make_shared<GCData>(type.getHeapType(), data);
+  //
+  // This consumes the input |data| entirely.
+  Literal makeGCData(Literals& data, Type type) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/wasm-interpreter.h]: Oh right, good idea. Done.

File: src/support/small_vector.h (hunk 1 above)
Comment (kripken) [reply src/support/small_vector.h:81]: Hmm, perhaps let's wait until there is a user? Otherwise adding it now would be dead/untested code.

Comment (tlively) [reply src/support/small_vector.h:81]: The problem is that if someone does try to use move assignment, the compiler will gracefully fall back to using copy assignment and the user will have no idea that they didn't get the move they wanted. We can prevent that by providing a move assignment operator eagerly, and the code should be simple enough that there shouldn't be any problems. (Alternatively we could provide a move assignment whose body is a WASM_UNREACHABLE.)

Comment (kripken) [reply src/support/small_vector.h:81]: Good point, that does sound risky. Added.

**Review by tlively — APPROVED** [review]

**Pull request merged**